File: rust/src/problems/p208.rs
pub struct Trie {
    next: Vec<Option<Box<Trie>>>,
    is_end: bool,
}

impl Trie {
    pub fn new() -> Self {
        Self {
            next: (0..26).map(|_| None).collect(),
            is_end: false,
        }
    }

    pub fn insert(&mut self, word: String) {
        let mut t = self;
        for &b in word.as_bytes() {
            let i = (b - b'a') as usize;
            t = t.next[i].get_or_insert_default();
        }
        t.is_end = true;
    }

    pub fn search(&self, word: String) -> bool {
        let mut t = self;
        for &b in word.as_bytes() {
            let i = (b - b'a') as usize;
            if let Some(trie) = &t.next[i] {
                t = trie;
            } else {
                return false;
            }
        }
        t.is_end
    }

    pub fn starts_with(&self, prefix: String) -> bool {
        let mut t = self;
        for &b in prefix.as_bytes() {
            let i = (b - b'a') as usize;
            if let Some(trie) = &t.next[i] {
                t = trie;
            } else {
                return false;
            }
        }
        true
    }
}

impl Default for Trie {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn case1() {
        let mut obj = Trie::new();
        obj.insert("apple".to_string());
        assert!(obj.search("apple".to_string()));
        assert!(!obj.search("app".to_string()));
        assert!(obj.starts_with("app".to_string()));
        obj.insert("app".to_string());
        assert!(obj.search("app".to_string()));
    }
}

File: cpp/p208.cpp
#include <cstdio>
#include <string>
#include <vector>

using namespace std;

class Trie {
    bool end;
    Trie *next[26];

public:
    Trie() : end(), next() {}

    void insert(string word) {
        Trie *t = this;
        for (auto c : word) {
            const auto i = c - 'a';
            if (!t->next[i]) {
                t->next[i] = new Trie();
            }
            t = t->next[i];
        }
        t->end = true;
    }

    bool search(string word) {
        Trie *t = this;
        for (auto c : word) {
            const auto i = c - 'a';
            if (!t->next[i]) {
                return false;
            }
            t = t->next[i];
        }
        return t->end == true;
    }

    bool startsWith(string prefix) {
        Trie *t = this;
        for (auto c : prefix) {
            const auto i = c - 'a';
            if (!t->next[i]) {
                return false;
            }
            t = t->next[i];
        }
        return true;
    }
};

int main() {
    vector<string> insert = {"", "insert", "apple"};
    vector<string> search = {"", "erqw", "insert", "apple"};
    vector<string> prefix = {"", "app", "appa"};

    auto t = Trie();
    for (auto &e : insert) {
        t.insert(e);
    }

    for (auto &e : search) {
        printf("%s search %d\n", e.data(), t.search(e));
    }

    for (auto &e : prefix) {
        printf("%s startsWith %d\n", e.data(), t.startsWith(e));
    }
    return 0;
}